toucan_connectors/workday/workday_connector.py

```diff
@@ -0,0 +1,116 @@
+        return (str(obj) for obj in [obj])
+    if isinstance(obj, types.GeneratorType):
+        return list(obj)
+    raise TypeError('Type %s not serializable' % type(obj))
+
+
+class WorkdayDataSource(ToucanDataSource):
+    service: str = Field(
+        None,
+        title='Name of the Workday API Service you want to use.',
+        description='Check "Public Web Services" page in Workday',
+        example='Human_Resources',
+    )
+    service_WSDL_URL: str = Field(
+        None,
+        title='URL of the WSDL of the Workday API Service you want to use.',
+        description='https://{Workday domain}/ccx/service/{tenant}/{service}/{version}?wsdl',
+        example='https://wd3-impl-services1.workday.com/ccx/service/umanis/Human_Resources/v33.2?wsdl',
+    )
+    operation: str = Field(
+        None,
+        title='Name of the Workday API Operation you want to call within the selected Service.',
+        description='Check https://community.workday.com/sites/default/files/file-hosting/productionapi/index.html',
+        example='Get_Workers',
+    )
+    request_parameters: dict = Field(
+        None,
+        title='Request References Parameter',
+        description='Check https://community.workday.com/sites/default/files/file-hosting/productionapi/index.html',
+        example='{Request_References : None, Request_Criteria : None, Response_Filter : {}, Response_Group : None}',
+    )
+    filter: str = FilterSchema
+
+
+class WorkdayConnector(ToucanConnector):
+    data_source_model: WorkdayDataSource
+
+    tenant: str = Field(
+        None,
+        title='Name of the Workday tenant',
+        description='Name of the Workday tenant',
+        example='my_tenant',
+    )
+    username: str = Field(
+        None,
+        title='Username for authentification',
+        description='Username for Credential authentification. Leave empty for Anonymous authentification.',
+        example='my_username',
+    )
+    password: str = Field(
+        None,
+        title='Password for authentification',
+        description='Password for Credential authentification',
+        example='$ecreTP4s$w0rd!',
+    )
+
+    def _retrieve_data(self, data_source: WorkdayDataSource) -> pd.DataFrame:
+        """ Only supports Anonymous or Credentials authentification """
+
+        if self.username:
+            auth = WsSecurityCredentialAuthentication(
+                '@'.join((self.username, self.tenant)), self.password
+            )
+        else:
+            auth = AnonymousAuthentication()
+
+        apis = {data_source.service: data_source.service_WSDL_URL}
+
+        client = workday.WorkdayClient(wsdls=apis, authentication=auth)
+
+        data = getattr(getattr(client, data_source.service), data_source.operation)(
+            **data_source.request_parameters
+        )
+
+        num_page = data.total_pages
+
+        data_json = json.loads(
+            json.dumps(zeep.helpers.serialize_object(data.data), default=json_serial)
```

Comment:
I still don't get it : 

```
>>> import json
>>> input = {}
>>> json.loads(json.dumps(input)) == input
True
```

Reply:
In our case, we do are using a custom JSONEncoder that overrides the default() method to serialize additional types. The zeep.helpers.serialize_object method doesn't return a serializable object.

Ok for you ?

Reply:
You are returning a DataFrame and not JSON text so why is it a problem to have dates and decimal ?

Reply:
Before returning a DataFrame but I need to transform a JSON with jq.

Here is the OrderedDict returned by zeep.helpers.serialize_object() :

> OrderedDict([('Absence_Input', [OrderedDict([('Absence_Input_Reference', OrderedDict([('ID', [OrderedDict([('_value
> _1', '08d4a6121b760154a3c7d6e1600c6646'), ('type', 'WID')])]), ('Descriptor', None)])), ('Absence_Input_Data', Orde
> redDict([('Absence_Input_ID', None), ('Batch_ID', 'Chargement Pris CP1 Golive Umanis'), ('Worker_Reference', Ordere
> dDict([('ID', [OrderedDict([('_value_1', '9836946ba6f4017d017f88c78d212b39'), ('type', 'WID')]), OrderedDict([('_va
> lue_1', 'E5042005'), ('type', 'Employee_ID')])]), ('Descriptor', None)])), ('Position_Reference', None), ('Absence_
> Component_Reference', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a6121b7601ba5f1109f95f0cfa2e'), ('type', 
> 'WID')]), OrderedDict([('_value_1', "FRA - Congés payés en cours d'acquisition (CP1) Time Off Reprise"), ('type', '
> Time_Off_Code')])]), ('Descriptor', None)])), ('Start_Date', datetime.date(2019, 1, 1)), ('End_Date', datetime.date
> (2019, 1, 31)), ('Reference_Date', datetime.date(2019, 1, 1)), ('Hours', Decimal('0')), ('Adjustment', True), ('Com
> ment', None)]))]), OrderedDict([('Absence_Input_Reference', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a61
> 21b76019c685b242a610cea54'), ('type', 'WID')])]), ('Descriptor', None)])), ('Absence_Input_Data', OrderedDict([('Ab
> sence_Input_ID', None), ('Batch_ID', 'Chargement Pris CP2 Golive Umanis'), ('Worker_Reference', OrderedDict([('ID',
>  [OrderedDict([('_value_1', '9836946ba6f401d1dbd602168e2192d4'), ('type', 'WID')]), OrderedDict([('_value_1', 'E508
> 0032'), ('type', 'Employee_ID')])]), ('Descriptor', None)])), ('Position_Reference', None), ('Absence_Component_Ref
> erence', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a6121b760197da744601600c1d2f'), ('type', 'WID')]), Ord
> eredDict([('_value_1', 'FRA - Congés Payés Acquis (CP2) Time Off Reprise'), ('type', 'Time_Off_Code')])]), ('Descri
> ptor', None)])), ('Start_Date', datetime.date(2019, 1, 1)), ('End_Date', datetime.date(2019, 1, 31)), ('Reference_D
> ate', datetime.date(2019, 1, 1)), ('Hours', Decimal('5')), ('Adjustment', True), ('Comment', None)]))])])])

And here is the error :

> ../../toucan_connectors/toucan_connector.py:162: in get_func_and_retry
>     return self.retry_decorator(func)(self, *args, **kwargs)
> ../../toucan_connectors/toucan_connector.py:236: in get_df
>     res = self._retrieve_data(data_source)
> ../../toucan_connectors/workday/workday_connector.py:142: in _retrieve_data
>     df = pd.DataFrame(use_jq_to_parse(zeep.helpers.serialize_object(data.data), data_source.filter))
> ../../toucan_connectors/workday/workday_connector.py:32: in use_jq_to_parse
>     return transform_with_jq(data, jq_filter)
> ../../toucan_connectors/common.py:273: in transform_with_jq
>     data = jq(jq_filter).transform(data, multiple_output=True)
> jq.pyx:178: in jq._Program.transform
>     ???
> jq.pyx:165: in jq._Program.input
>     ???
> /usr/lib/python3.6/json/__init__.py:231: in dumps
>     return _default_encoder.encode(obj)
> /usr/lib/python3.6/json/encoder.py:199: in encode
>     chunks = self.iterencode(o, _one_shot=True)
> /usr/lib/python3.6/json/encoder.py:257: in iterencode
>     return _iterencode(o, 0)
> 
> TypeError: Object of type 'Decimal' is not JSON serializable

Reply:
I understand the problem, my point is can we not have a simple / less expensive way to get the type we want : I see for example that the `zeep.helpers.serialize_object` takes an extra `target_cls` argument. cf. https://docs.python-zeep.org/en/master/helpers.html#zeep.helpers.serialize_object have you looked into this ?